stan/math/prim/mat/prob/neg_binomial_2_log_glm_lpmf.hpp

```diff
@@ -95,107 +124,94 @@ neg_binomial_2_log_glm_lpmf(const T_y& y, const T_x& x, const T_alpha& alpha,
+  const auto& alpha_val = value_of_rec(alpha);
+  const auto& phi_val = value_of_rec(phi);
 
-  Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
-  {
-    scalar_seq_view<T_beta> beta_vec(beta);
-    for (size_t m = 0; m < M; ++m) {
-      beta_dbl[m] = value_of(beta_vec[m]);
-    }
-  }
+  const auto& y_val_vec = as_column_vector_or_scalar(y_val);
+  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto& alpha_val_vec = as_column_vector_or_scalar(alpha_val);
+  const auto& phi_val_vec = as_column_vector_or_scalar(phi_val);
 
-  Array<T_partials_return, Dynamic, 1> theta_dbl
-      = (value_of(x) * beta_dbl).array();
-  scalar_seq_view<T_alpha> alpha_vec(alpha);
+  const auto& y_arr = as_array_or_scalar(y_val_vec);
+  const auto& phi_arr = as_array_or_scalar(phi_val_vec);
+
+  Array<T_partials_return, Dynamic, 1> theta = value_of(x) * beta_val_vec;
+  theta += as_array_or_scalar(alpha_val_vec);
   for (size_t n = 0; n < N; ++n)
-    theta_dbl[n] += value_of(alpha_vec[n]);
-  Array<T_partials_return, Dynamic, 1> log_phi = phi_arr.log();
-  Array<T_partials_return, Dynamic, 1> logsumexp_eta_logphi
-      = theta_dbl.binaryExpr(log_phi, [](const T_partials_return& xx,
-                                         const T_partials_return& yy) {
-          return log_sum_exp(xx, yy);
-        });
-  Array<T_partials_return, Dynamic, 1> y_plus_phi = y_arr + phi_arr;
+    check_finite(function, "Matrix of independent variables", theta[n]);
+  T_precision_val log_phi = log(phi_arr);
+  Array<T_partials_return, Dynamic, 1> logsumexp_theta_logphi
+      = (theta > log_phi)
+            .select(theta + Eigen::log1p(Eigen::exp(log_phi - theta)),
+                    log_phi + Eigen::log1p(Eigen::exp(theta - log_phi)));
+
+  T_sum_val y_plus_phi = y_arr + phi_arr;
 
   // Compute the log-density.
   if (include_summand<propto>::value) {
-    logp -= (y_arr + Array<double, Dynamic, 1>::Ones(N, 1))
-                .unaryExpr(
-                    [](const T_partials_return& xx) { return lgamma(xx); })
-                .sum();
+    logp -= sum(lgamma(y_arr + 1));
   }
   if (include_summand<propto, T_precision>::value) {
-    for (size_t n = 0; n < N; ++n)
-      logp += multiply_log(phi_arr[n], phi_arr[n]) - lgamma(phi_arr[n]);
+    if (is_vector<T_precision>::value) {
+      scalar_seq_view<decltype(phi_val)> phi_vec(phi_val);
+      for (size_t n = 0; n < N; ++n)
+        logp += multiply_log(phi_vec[n], phi_vec[n]) - lgamma(phi_vec[n]);
+    } else {
+      logp += N
+              * (multiply_log(as_scalar(phi_val), as_scalar(phi_val))
+                 - lgamma(as_scalar(phi_val)));
+    }
   }
   if (include_summand<propto, T_x, T_alpha, T_beta, T_precision>::value)
-    logp -= (y_plus_phi * logsumexp_eta_logphi).sum();
+    logp -= (y_plus_phi * logsumexp_theta_logphi).sum();
   if (include_summand<propto, T_x, T_alpha, T_beta>::value)
-    logp += (y_arr * theta_dbl).sum();
+    logp += (y_arr * theta).sum();
```

Comment:
I hesitate to comment on a lot of this, but we'd generally prefer to use `stan::math::sum` as it's more efficient for autdiff types (it consolidates expression graphs).  This is only going to matter for higher-order autodiff, which we aren't even using now.  But if we leave these in as `sum()`, we'll probably have to change them later for efficiency.  Anyway, no big deal leaving them in.